core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ChangeMoveProvider.java

```diff
@@ -0,0 +1,47 @@
+package ai.timefold.solver.core.impl.move.streams.maybeapi.generic.provider;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.impl.move.streams.DefaultMoveStreamFactory;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.ChangeMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProducer;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProvider;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveStreamFactory;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.SolutionViewTriPredicate;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PlanningVariableMetaModel;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public final class ChangeMoveProvider<Solution_, Entity_, Value_>
+        implements MoveProvider<Solution_> {
+
+    private final PlanningVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel;
+    private final SolutionViewTriPredicate<Solution_, Entity_, @Nullable Value_> entityValueFilter;
+
+    public ChangeMoveProvider(PlanningVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel) {
+        this.variableMetaModel = Objects.requireNonNull(variableMetaModel);
+        this.entityValueFilter = (solutionView, entity, value) -> {
+            Value_ oldValue = solutionView.getValue(variableMetaModel, entity);
+            if (Objects.equals(oldValue, value)) {
+                return false;
+            }
+            return solutionView.isValueInRange(variableMetaModel, entity, value);
```

Comment:
I understand the approach, but I'm thinking if it might include some unnecessary overhead by performing a lookup every time.

Reply:
Since we can determine whether the range descriptor is entity-independent, I suggest avoiding the execution of the filter in these cases. This approach would maintain the logic of ValueRangeState in relation to the lookup logic, while also eliminating unnecessary calls for value ranges placed in solutions.

Reply:
Resolving in favor of a similar comment further down.